doc/explanation/mental-model.rst

```diff
@@ -0,0 +1,175 @@
+The Dune Mental Model
+=====================
+
+It is not strictly necessary to understand Dune's underlying model to use it;
+but knowing how it works under the hood will help understand some errors and
+what's possible with Dune.
```

Comment:
I think it also makes writing rules easier if you know what you're doing instead of (I assume many like me started out just copying boilerplate from somewhere else).